pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,298 @@
+	ctx   context.Context
+	errCh chan error // Signalled to propagate terminal error the consumer.
+	input chan eventOrError
+}
+
+// Recv implements rangeFeedEventProducer interface.
+func (c *channelRangeFeedEventProducer) Recv() (*roachpb.RangeFeedEvent, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, c.ctx.Err()
+	case err := <-c.errCh:
+		return nil, err
+	case e := <-c.input:
+		return e.event, e.err
+	}
+}
+
+var _ roachpb.RangeFeedEventProducer = (*channelRangeFeedEventProducer)(nil)
+
+// startMuxRangeFeed begins the execution of rangefeed for the specified
+// RangeFeedRequest.
+func (m *rangefeedMuxer) startMuxRangeFeed(
+	ctx context.Context, client rpc.RestrictedInternalClient, req *roachpb.RangeFeedRequest,
+) (_ roachpb.RangeFeedEventProducer, cleanup func(), _ error) {
+	producer, rpcClient, streamID, cleanup, err := m.connect(ctx, client, req.Replica.NodeID)
+	if err != nil {
+		return nil, cleanup, err
+	}
+	req.StreamID = streamID
+	if err := rpcClient.Send(req); err != nil {
+		return nil, cleanup, err
+	}
+	return producer, cleanup, nil
+}
+
+// connect establishes MuxRangeFeed connection for the specified node. returns
+// event producer, RPC client to send requests on, the streamID which should be
+// used when sending request, and a cleanup function. Cleanup function never
+// nil, and must always be invoked, even if error is returned.
+func (m *rangefeedMuxer) connect(
+	ctx context.Context, client rpc.RestrictedInternalClient, nodeID roachpb.NodeID,
+) (
+	_ roachpb.RangeFeedEventProducer,
+	_ roachpb.Internal_MuxRangeFeedClient,
+	streamID int64,
+	cleanup func(),
+	_ error,
+) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	streamID = m.mu.nextStreamID
+	m.mu.nextStreamID++
+
+	cleanup = func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		delete(m.mu.producers, streamID)
+		muxState := m.mu.muxClients[nodeID]
+		if muxState != nil {
+			muxState.streams.Remove(int(streamID))
+			if muxState.streams.Len() == 0 {
+				// This node no longer has any active streams.
+				// Delete node from the muxClient list, and gracefully
+				// shutdown consumer go routine.
+				delete(m.mu.muxClients, nodeID)
+				muxState.cancel()
+			}
+		}
+	}
+
+	if m.mu.terminalErr != nil {
+		return nil, nil, streamID, cleanup, m.mu.terminalErr
+	}
+
+	// Lookup or establish MuxRangeFeed for this node.
+	var found bool
+	ms, found := m.mu.muxClients[nodeID]
+
+	if !found {
+		ctx, cancel := context.WithCancel(ctx)
+		stream, err := client.MuxRangeFeed(ctx)
+		if err != nil {
+			cancel()
+			return nil, nil, streamID, cleanup, err
+		}
+
+		ms = &muxClientState{client: stream, cancel: cancel}
+		m.mu.muxClients[nodeID] = ms
+		m.g.GoCtx(func(ctx context.Context) error {
+			defer cancel()
+			return m.receiveEventsFromNode(ctx, nodeID, stream)
+		})
+	}
+
+	// Start RangeFeed for this request.
+	ms.streams.Add(int(streamID))
+
+	producer := &channelRangeFeedEventProducer{
+		ctx:   ctx,
+		errCh: make(chan error, 1),
```

Comment:
Bit unexpected to see an error chan and eventOrError next to it, add a comment.

Reply:
Fixed.